lib/internal/per_context/safethenable.js

```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+const {
+  PromisePrototypeCatch,
+  PromisePrototypeThen,
+  ReflectApply,
+  SafeWeakMap,
+} = primordials;
+
+const { isPromise } = require('internal/util/types');
+
+const cache = new SafeWeakMap();
+
+// `SafeThenable` should be used when dealing with user provided `Promise`-like
+// instances. It provides two methods `.then` and `.catch`.
+// Wrapping uses of `SafeThenable` in `try`/`catch` may be useful if the
+// accessing `.then` property of the provided object throws.
+class SafeThenable {
+  #is_promise;
+  #makeUnsafeCalls;
+  #target;
+  #cachedThen;
+  #hasAlreadyAccessedThen;
+
+  constructor(thenable, makeUnsafeCalls = false) {
+    this.#target = thenable;
+    this.#is_promise = isPromise(thenable);
+    this.#makeUnsafeCalls = makeUnsafeCalls;
+  }
+
+  get #then() {
+    // Handle Promises/A+ spec, `then` could be a getter
+    // that throws on second access.
+    if (this.#hasAlreadyAccessedThen === undefined) {
+      this.#hasAlreadyAccessedThen = cache.has(this.#target);
+      if (this.#hasAlreadyAccessedThen) {
+        this.#cachedThen = cache.get(this.#target);
+      }
+    }
+    if (!this.#hasAlreadyAccessedThen) {
+      this.#cachedThen = this.#target?.then;
+      this.#hasAlreadyAccessedThen = true;
+      if (typeof this.#target === 'object' && this.#target !== null) {
+        cache.set(this.#target, this.#cachedThen);
+      }
+    }
+
+    return this.#cachedThen;
+  }
+
+  get isThenable() {
+    return this.#is_promise || typeof this.#then === 'function';
+  }
+
+  catch(onError) {
+    return this.#is_promise ?
+      PromisePrototypeCatch(this.#target, onError) :
+      this.then(undefined, onError);
+  }
+
+  then(...args) {
```

Comment:
This is terrifying, even if we ignore things like "the returned thenable is missing finally" or "if you do `.constructor you don't get all the promise statics. This is still terrifying ^^

I am -0.5, if others feel strongly then _maybe_

Reply:
This is intended to be used with objects for which only `.then` method is called. It's just a wrapper for the pattern that shows several times in core:

```js
const then = obj?.then;
if(typeof then === function) {
  FunctionPrototypeCall(then, obj, onSuccess, onError);
}
```

I think `.finally` is a bit off-topic, we should always prefer `PromisePrototypeFinally` anyway.